lib/chef-dk/policyfile/undo_record.rb

```diff
@@ -0,0 +1,135 @@
+#
+# Copyright:: Copyright (c) 2015 Chef Software Inc.
+# License:: Apache License, Version 2.0
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+require 'chef-dk/exceptions'
+
+module ChefDK
+  module Policyfile
+
+    class UndoRecord
+
+      PolicyGroupRestoreData = Struct.new(:policy_name, :policy_group, :data) do
+
+        def load(data)
+          self.policy_name = data["policy_name"]
+          self.policy_group = data["policy_group"]
+          self.data = data["data"]
+          self
+        end
+
+        def for_serialization
+          {
+            "policy_name" => policy_name,
+            "policy_group" => policy_group,
+            "data" => data
+          }
+        end
+
+      end
+
+      attr_reader :policy_groups
+
+      attr_reader :policy_revisions
+
+      def initialize
+        reset!
+      end
+
+      def ==(other)
+        other.kind_of?(UndoRecord) &&
+          other.policy_groups == policy_groups &&
+          other.policy_revisions == policy_revisions
+      end
+
+      def add_policy_group(name)
+        @policy_groups << name
+      end
+
+      def add_policy_revision(policy_name, policy_group, data)
+        @policy_revisions << PolicyGroupRestoreData.new(policy_name, policy_group, data)
+      end
+
+      def load(data)
+        reset!
+
+        unless data.kind_of?(Hash)
+          raise InvalidUndoRecord, "Undo data is incorrectly formatted. Must be a Hash, got '#{data}'."
```

Comment:
I want to display `data` because it might help you figure out what went wrong. But now that I think of it, this should probably be `data.inspect`